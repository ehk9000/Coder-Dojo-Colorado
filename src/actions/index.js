export const addUser = (user) => ({
  type:'ADD_USER',
  user
});

export const isFetching = (bool) => ({
  type: 'IS_FETCHING',
  isFetching: bool
});