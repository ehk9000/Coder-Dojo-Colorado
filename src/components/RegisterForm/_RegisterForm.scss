.RegisterForm {
  margin: 0 auto;
  width:50%;
  @include flex-column;
  
  label {
    @include flex-column;
  }
  
  .signin-btn {
    @include form-submit;
  }

  #full-name-input,
  #username-input,
  #password-input,
  #reentered-password-input {
    @include form-input-style
  }


}