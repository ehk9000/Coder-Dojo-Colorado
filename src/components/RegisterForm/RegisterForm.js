import React, { Component } from 'react';
import { PropTypes } from 'prop-types';
import { connect } from 'react-redux';
import * as actions from '../../actions';
import ContactForm from '../ContactForm/ContactForm';

export class RegisterForm extends Component {
  state = {
    fullName: '',
    username: '',
    password: '',
    reEnteredPassword: '',
    displayContactForm: false
  }

  handleChange = (e) => {
    const { name, value } = e.target;

    this.setState({
      [name]: value
    });
  }

  handleRegister = (e) => {
    const { name } = e.target;
    const { 
      username, 
      password, 
      fullName, 
      reEnteredPassword 
    } = this.state;
    const { addUser } = this.props;
    const newUser = {
      username,
      password,
      fullName
    };
    let displayContactForm;

    e.preventDefault();

    if (name === 'student') {
      addUser(newUser);
    } else {
      this.setState({ displayContactForm: true });
      displayContactForm = <ContactForm {...this.state} />
    }
    return displayContactForm;
  }

  assignContactForm = () => {
    let assignContact;
    const { displayContactForm } = this.state;
    if (displayContactForm) {
      assignContact = <ContactForm {...this.state} />;
    }
    return assignContact;
  }

  render() { 
    const displayContactForm = this.assignContactForm();
    return (
      <section>
        <form 
          className="RegisterForm"
          onSubmit={this.handleRegister}
        >
          <h2>Register an Account</h2>
          <label htmlFor="full-name-input">
            Full Name
            <input 
              id="full-name-input"
              type="text" 
              name="fullName"
              onChange={this.handleChange}
            />
          </label>
          <label htmlFor="username-input">
            Username
            <input 
              id="username-input"
              type="text"
              name="username"
              onChange={this.handleChange}
            />
          </label>
          <label htmlFor="password-input">
            Password
            <input 
              id="password-input"
              type="password"
              name="password"
              onChange={this.handleChange}
            />
          </label>
          <label htmlFor="reentered-password-input">
            Re-enter Password
            <input 
              id="reentered-password-input"
              type="password"
              name="reEnteredPassword"
              onChange={this.handleChange}
            />
          </label>
          <button type="button" className="signin-btn" value="submit" name="student" onClick={this.handleRegister}>I am a student over 13</button>
          <button type="button" className="signin-btn guardian-btn" name="guardian" onClick={this.handleRegister}>I am a guardian</button>
        </form>
        {displayContactForm}
      </section>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  addUser: user => dispatch(actions.addUser(user))
});

export default connect(null, mapDispatchToProps)(RegisterForm);

RegisterForm.propTypes = {
  addUser: PropTypes.func
};

RegisterForm.defaultProps = {
  addUser: () => {}
};
