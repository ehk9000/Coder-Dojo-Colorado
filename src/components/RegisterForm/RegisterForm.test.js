import React from 'react';
import { shallow } from 'enzyme';
import { RegisterForm, mapDispatchToProps } from './RegisterForm';
import * as actions from '../../actions';

jest.mock('../../actions');


describe('RegisterForm', () => {
  let wrapper;
  const mockDispatch = jest.fn();


  beforeEach(() => {
    wrapper = shallow(<RegisterForm />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state on input change', () => {
    expect(wrapper.state()).toEqual({
      fullName: '',
      username: '',
      password: '',
      reEnteredPassword: '',
      displayContactForm: false,
      displayAgeForm: false
    }); 

    const mockEvent = {
      target: { value: 'ehk', name: 'username' }
    };

    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('username')).toEqual('ehk');
  });

  it('should dispatch addUser', () => {
    const mockUser = { username: 'ehk' };
    const mockAction = actions.addUser(mockUser);

    mapDispatchToProps(mockDispatch).addUser(mockUser);

    expect(mockDispatch).toHaveBeenCalledWith(mockAction);  
  });
});
