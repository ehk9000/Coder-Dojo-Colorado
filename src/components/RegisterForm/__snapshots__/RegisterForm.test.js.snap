// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm should match snapshot 1`] = `
<section>
  <form
    className="RegisterForm"
    onSubmit={[Function]}
  >
    <h2>
      Register an Account
    </h2>
    <label
      htmlFor="full-name-input"
    >
      Full Name
      <input
        id="full-name-input"
        name="fullName"
        onChange={[Function]}
        type="text"
      />
    </label>
    <label
      htmlFor="username-input"
    >
      Username
      <input
        id="username-input"
        name="username"
        onChange={[Function]}
        type="text"
      />
    </label>
    <label
      htmlFor="password-input"
    >
      Password
      <input
        id="password-input"
        name="password"
        onChange={[Function]}
        type="password"
      />
    </label>
    <label
      htmlFor="reentered-password-input"
    >
      Re-enter Password
      <input
        id="reentered-password-input"
        name="reEnteredPassword"
        onChange={[Function]}
        type="password"
      />
    </label>
    <div
      className="error-msg"
    />
    <button
      className="signin-btn"
      name="student"
      onClick={[Function]}
      type="button"
      value="submit"
    >
      I am a student over 13
    </button>
    <button
      className="signin-btn guardian-btn"
      name="guardian"
      onClick={[Function]}
      type="button"
    >
      I am a guardian
    </button>
  </form>
</section>
`;
